rllib/env/remote_base_env.py

```diff
@@ -49,71 +55,88 @@ def __init__(self,
         self.poll_timeout = remote_env_batch_wait_ms / 1000
+        # Already existing env objects (generated by the RolloutWorker).
+        self.existing_envs = existing_envs or []
+        self.worker = worker
+
+        # Whether the given `make_env` callable already returns ray.remote
+        # objects or not.
+        self.make_env_creates_actors = False
+
+        self._observation_space = None
+        self._action_space = None
 
         # List of ray actor handles (each handle points to one @ray.remote
         # sub-environment).
         self.actors: Optional[List[ray.actor.ActorHandle]] = None
-        self._observation_space = None
-        self._action_space = None
+
+        # `self.make_env` already produces Actors: Use it directly.
+        if len(self.existing_envs) > 0 and isinstance(self.existing_envs[0],
+                                                      ray.actor.ActorHandle):
+            self.make_env_creates_actors = True
+            self.actors = []
+            while len(self.actors) < self.num_envs:
+                sub_env = self.make_env(len(self.actors))
+                if self.worker is not None:
+                    self.worker.callbacks.on_sub_environment_created(
+                        worker=self.worker,
+                        sub_environment=sub_env,
+                        env_context=self.worker.env_context.
+                        copy_with_overrides(vector_index=len(self.actors)))
+                self.actors.append(sub_env)
+        # `self.make_env` produces gym.Envs (or children thereof, such
+        # as MultiAgentEnv): Need to auto-wrap it here. The problem with
+        # this is that custom methods wil get lost. If you would like to
+        # keep your custom methods in your envs, you should provide the
+        # env class directly in your config (w/o tune.register_env()),
+        # such that your class will directly be made a @ray.remote
+        # (w/o the wrapping via `_Remote[Multi|Single]AgentEnv`).
+        else:
+
+            def make_remote_env(i):
+                logger.info("Launching env {} in remote actor".format(i))
+                if self.multiagent:
+                    sub_env = _RemoteMultiAgentEnv.remote(self.make_env, i)
+                else:
+                    sub_env = _RemoteSingleAgentEnv.remote(self.make_env, i)
+
+                if self.worker is not None:
+                    self.worker.callbacks.on_sub_environment_created(
+                        worker=self.worker,
+                        sub_environment=sub_env,
+                        env_context=self.worker.env_context.
+                        copy_with_overrides(vector_index=i))
+                return sub_env
+
+            self.actors = [make_remote_env(i) for i in range(self.num_envs)]
+            self._observation_space = ray.get(
+                self.actors[0].observation_space.remote())
+            self._action_space = ray.get(self.actors[0].action_space.remote())
```

Comment:
I feel like these 2 calls don't need to be remote(). probably happens faster if we just call them locally on the driver node?

Reply:
Not sure how you would achieve that. Unless we had at least one `existing_envs` (which in this else block would not be an actor, but an actual local env object).
If we don't have `existing_envs`, we would not know what the make_env function returns (because we already create the actual sub-envs inside our actor wrapper: `_RemoteMultiAgentEnv` or `_RemoteSingleAgentEnv`). So then we have to do the remote call. But I can add a `existing_envs` check to avoid this in case we do have `len(existing_envs) > 0`.

Reply:
All fixed.

Reply:
makes sense, thanks